scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/license/LicenseSchedulingPolicy.java

```diff
@@ -116,26 +198,61 @@ public boolean isTaskExecutable(NodeSet selectedNodes, EligibleTaskDescriptor ta
 
     }
 
-    private void removeFinishedTasks(LinkedBlockingQueue<EligibleTaskDescriptor> eligibleTasksDescriptorsLicense) {
+    private static void removeFinishedJobs(LinkedBlockingQueue<JobDescriptorImpl> eligibleJobsDescriptorsLicense) {
+        Iterator<JobDescriptorImpl> iter = eligibleJobsDescriptorsLicense.iterator();
+        JobDescriptorImpl currentJob;
+        while (iter.hasNext()) {
+            currentJob = iter.next();
+
+            if (!currentJob.getInternal().getStatus().isJobAlive()) {
+                iter.remove();
+            }
+        }
+    }
+
+    private static boolean canGetJobLicense(String currentRequiredLicense) {
+
+        // If the required software is not specified in the license properties file
+        if (!eligibleJobsDescriptorsLicenses.containsKey(currentRequiredLicense))
+            return false;
+
+        // if there are still remaining licenses return true
+        LinkedBlockingQueue<JobDescriptorImpl> eligibleJobsDescriptorsLicense = eligibleJobsDescriptorsLicenses.get(currentRequiredLicense);
+        if (eligibleJobsDescriptorsLicense.remainingCapacity() > 0) {
+            return true;
+        }
+        // otherwise remove all terminated jobs and check again if there are remaining licenses
+        else {
+            removeFinishedJobs(eligibleJobsDescriptorsLicense);
```

Comment:
this is called every time a licence is checked, and it iterates by itself on all jobs. 

I think it should be factorized to be called only once when getOrderedTasks is called.

Reply:
ok

Reply:
Finally following our discussion, we will keep it.